### PR TITLE
feat: Accept contentType params for DID URL Resolution [DEV-4722]

### PR DESCRIPTION
**Pull request by DaevMithran**

### CODE REVIEW

**Comment by ankurdotb** [comment]
Task linked: [DEV-4722 Apply Accept header logic for DID URL Dereferencing](https://app.clickup.com/t/6600954/DEV-4722)

File: services/resource/echo_handlers.go [hunk 1]
```diff
@@ -22,6 +23,13 @@ import (
 //	@Failure		501			{object}	types.IdentityError
 //	@Router			/{did}/resources/{resourceId} [get]
 func ResourceDataEchoHandler(c echo.Context) error {
+	// Get Accept header
+	requestedContentTypeParam := services.ExtractMediaTypeParams(c.Request().Header.Get(echo.HeaderAccept))
```
[the remark this answers is not in the document]
Comment (sownak) [reply services/resource/echo_handlers.go:27]: `Accept: text/xml;q=0.8, application/json;q=0.9`

File: services/helpers.go [hunk 2]
```diff
@@ -39,3 +39,27 @@ func PrepareQueries(c echo.Context) (rawQuery string, flag *string) {
 func GetDidParam(c echo.Context) (string, error) {
 	return url.QueryUnescape(c.Param("did"))
 }
+
+func ExtractMediaTypeParams(mediaType string) map[string]string {
```
Comment (ankurdotb) [inline services/helpers.go:43]: As far as I understand, this is just stripping out the `q` values without taking it into consideration for priority order?

File: services/resource/echo_handlers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ankurdotb) [reply services/resource/echo_handlers.go:27]: @DaevMithran This is **in** the task description :) Please follow the link on it for explanation and examples.

File: types/constants.go [hunk 3]
```diff
@@ -3,10 +3,11 @@ package types
...
 const (
-	DIDJSON   ContentType = "application/did+json"
-	DIDJSONLD ContentType = "application/did+ld+json"
-	JSONLD    ContentType = "application/ld+json"
-	JSON      ContentType = "application/json"
+	DIDJSON    ContentType = "application/did+json"
+	DIDJSONLD  ContentType = "application/did+ld+json"
+	JSONLD     ContentType = "application/ld+json"
+	JSON       ContentType = "application/json"
+	W3IDDIDURL string      = "https://w3id.org/did-url-dereferencing"
```
Comment (ankurdotb) [inline types/constants.go:10]: There's two different headers/profiles but only one listed here, so I think this hasn't been accounted for at all. This is `profile="https://w3id.org/did-resolution"` described in the task, which is different then URL dereferencing for resources.

File: types/resource_dereferencing.go [hunk 4]
```diff
@@ -1,10 +1,10 @@
 package types
 
 type ResourceDereferencing struct {
-	Context               string                     `json:"@context,omitempty" example:"https://w3id.org/did-resolution/v1"`
-	DereferencingMetadata DereferencingMetadata      `json:"dereferencingMetadata"`
-	ContentStream         ContentStreamI             `json:"contentStream"`
-	Metadata              ResolutionResourceMetadata `json:"contentMetadata"`
+	Context               string                `json:"@context,omitempty" example:"https://w3id.org/did-resolution/v1"`
```
Comment (ankurdotb) [inline types/resource_dereferencing.go:4]: Note that this header should only be inserted if the Accept header has some combination of LD+JSON (i.e., they have requested JSON-LD).

**Review by ankurdotb — CHANGES_REQUESTED** [review]

File: types/constants.go (hunk 3 above)
Comment (DaevMithran) [reply types/constants.go:10]: @ankurdotb That is DID Resolution work is in a separate PR by sownak. This is specific to DID URL Resolution

File: services/helpers.go (hunk 2 above)
Comment (DaevMithran) [reply services/helpers.go]: Added Priority handling in accept headers now

File: types/resource_dereferencing.go (hunk 4 above)
Comment (DaevMithran) [reply types/resource_dereferencing.go:4]: Yep. This is handled already

**Review by ankurdotb — CHANGES_REQUESTED** [review]
Maybe the functionality of the overall task is meant to be delivered across different PR’s and I’m struggling to see the big picture, but AFAIK I don’t see anything in here that’s handling or testing what happens when the correct header/profile for URL dereferencing is supplied, but the resource is not JSON/text. (At the very least, it’s not in the tests.) Might be a function of no resources being created that are image or non-text/JSON perhaps?

"As-is" situations where only the resource is returned I'm less worried about.

File: types/resource_data.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	if err != nil {
+		log.Info().Msg("Failed to marshal resource")
+		return []byte{}
+	}
+	return bytes
+}
+
+func (e *ResourceData) MarshalJSON() ([]byte, error) {
+	return json.Marshal(e.Content)
+}
```
[the remark this answers is not in the document]
Comment (DaevMithran) [reply types/resource_data.go:36]: This function overrides the default json marshalling behaviour of a struct. This helps in printing the content top level instead under a repetitive key content.

File: types/resource_data.go
[the remark this answers is not in the document]
Comment (DaevMithran) [reply types/resource_data.go:1]: By default bytes are converted to base64, Introducing this struct helps in unmarshalling bytes to json and returning it in the response 

Usage: https://github.com/cheqd/did-resolver/blob/DEV-4722/services/resource_dereference_service.go#L102

File: tests/integration/rest/resource/data/positive_test.go
[the remark this answers is not in the document]
Comment (DaevMithran) [reply tests/integration/rest/resource/data/positive_test.go:1]: They are added under data_with_metadata folder

File: services/resource_dereference_service.go [hunk 6]
```diff
@@ -79,3 +78,31 @@ func (rds ResourceService) DereferenceResourceData(did string, resourceId string
...
+	dereferenceMetadata := types.NewDereferencingMetadata(did, contentType, "")
+
+	resource, err := rds.ledgerService.QueryResource(did, strings.ToLower(resourceId))
+	if err != nil {
+		err.ContentType = contentType
+		return nil, err
+	}
+
+	var context string
+	if contentType == types.DIDJSONLD || contentType == types.JSONLD {
```
[the remark this answers is not in the document]
Comment (DaevMithran) [reply services/resource_dereference_service.go:92]: we should populdate context if it's jsonLD right. This condition is for that.

File: tests/integration/rest/resource/data/positive_test.go
[the remark this answers is not in the document]
Comment (ankurdotb) [reply tests/integration/rest/resource/data/positive_test.go:1]: Hence the "EDIT". The follow-up is about adding Chrome-like Accept header to test the "default" case of loading in a browser.

File: services/helpers.go [hunk 7]
```diff
@@ -39,3 +40,48 @@ func PrepareQueries(c echo.Context) (rawQuery string, flag *string) {
 func GetDidParam(c echo.Context) (string, error) {
 	return url.QueryUnescape(c.Param("did"))
 }
+
+func GetHighestPriorityHeaderAndParams(acceptHeader string) (string, map[string]string) {
```
[the remark this answers is not in the document]
Comment (DaevMithran) [reply services/helpers.go]: Using go-accept-headers which covers all the edge cases

File: tests/integration/rest/resource/data_with_metadata/positive_test.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	Entry(
+		"can get resource with metadata with an existent DID, with multiple header and q values",
+		utils.PositiveTestCase{
+			DidURL: fmt.Sprintf(
+				"http://%s/1.0/identifiers/%s/resources/%s",
+				testconstants.TestHostAddress,
+				testconstants.UUIDStyleTestnetDid,
+				testconstants.UUIDStyleTestnetDidResourceId,
+			),
+			ResolutionType:       string(types.JSONLD) + ";profile=" + string(types.W3IDDIDURL) + ";q=1.0,application/json;q=0.9,image/png;q=0.7",
```
[the remark this answers is not in the document]
Comment (DaevMithran) [reply tests/integration/rest/resource/data_with_metadata/positive_test.go]: This is done

File: tests/integration/rest/resource/data_with_metadata/positive_test.go
[the remark this answers is not in the document]
Comment (DaevMithran) [reply tests/integration/rest/resource/data_with_metadata/positive_test.go:1]: Added validation in data_with_metadata to check if base64 or json

**Comment by DaevMithran** [comment]
Duplicate of https://github.com/cheqd/did-resolver/pull/340
